tutorials/publish_subscribe/c/home_automation_subscriber.c

```diff
@@ -0,0 +1,182 @@
+            &info_seq,
+            DDS_LENGTH_UNLIMITED,
+            DDS_ANY_SAMPLE_STATE,
+            DDS_ANY_VIEW_STATE,
+            DDS_ANY_INSTANCE_STATE);
+    if (retcode == DDS_RETCODE_NO_DATA) {
+        return;
+    } else if (retcode != DDS_RETCODE_OK) {
+        fprintf(stderr, "take error %d\n", retcode);
+        return;
+    }
+
+    for (i = 0; i < DeviceStatusSeq_get_length(&data_seq); ++i) {
+        if (DDS_SampleInfoSeq_get_reference(&info_seq, i)->valid_data) {
+            device_status = DeviceStatusSeq_get_reference(&data_seq, i);
+            if (device_status->is_open) {
+                printf("WARNING: %s in %s is open!\n",
+                        device_status->sensor_name,
+                        device_status->room_name);
+            }
+        }
+    }
+
+    retcode = DeviceStatusDataReader_return_loan(
+            DeviceStatus_reader,
+            &data_seq,
+            &info_seq);
+    if (retcode != DDS_RETCODE_OK) {
+        fprintf(stderr, "return loan error %d\n", retcode);
+    }
+}
+
+int monitor_sensor(void)
+{
+
+    DDS_DomainParticipant *participant = NULL;
+    DDS_Subscriber *subscriber = NULL;
+    DDS_Topic *topic = NULL;
+    struct DDS_DataReaderListener reader_listener =
+            DDS_DataReaderListener_INITIALIZER;
+    DDS_DataReader *reader = NULL;
+    const char *type_name = NULL;
+    DDS_ReturnCode_t retcode;
+    struct DDS_Duration_t poll_period = {2,0};
+    int i;
+
+    participant = DDS_DomainParticipantFactory_create_participant(
+            DDS_TheParticipantFactory,
+            0,
+            &DDS_PARTICIPANT_QOS_DEFAULT,
+            NULL /* listener */,
+            DDS_STATUS_MASK_NONE);
+
+    if (participant == NULL) {
+        fprintf(stderr, "create_participant error\n");
+        subscriber_shutdown(participant);
+        return -1;
+    }
+
+    subscriber = DDS_DomainParticipant_create_subscriber(
+            participant,
+            &DDS_SUBSCRIBER_QOS_DEFAULT,
+            NULL /* listener */,
+            DDS_STATUS_MASK_NONE);
+
+    if (subscriber == NULL) {
+        fprintf(stderr, "create_subscriber error\n");
+        subscriber_shutdown(participant);
+        return -1;
+    }
+
+    type_name = DeviceStatusTypeSupport_get_type_name();
+    retcode = DeviceStatusTypeSupport_register_type(
+            participant,
+            type_name);
+
+    if (retcode != DDS_RETCODE_OK) {
+        fprintf(stderr, "register_type error %d\n", retcode);
+        subscriber_shutdown(participant);
+        return -1;
+    }
+
+    topic = DDS_DomainParticipant_create_topic(
+            participant,
+            "WindowStatus",
+            type_name,
+            &DDS_TOPIC_QOS_DEFAULT,
+            NULL /* listener */,
+            DDS_STATUS_MASK_NONE);
+
+    if (topic == NULL) {
+        fprintf(stderr, "create_topic error\n");
+        subscriber_shutdown(participant);
+        return -1;
+    }
+
+    reader_listener.on_data_available =
```

Comment:
let's use `DDS_SampleProcessor` in C as well.

Reply:
Didn't know Sample Processor exists for C. I will push the changes with it :+1: